landing_page: Restructure top-level navigation.

eeshangarg opened the pull request:
@timabbott, @amanagr, @alya: This took way way longer than I expected. But this is how I am hoping to implement dropdowns for the top-level navigation.
<img width="1345" alt="2021-06-23" src="https://user-images.githubusercontent.com/7251823/123168266-4ed63500-d452-11eb-9fa7-f29e09f06763.png">

timabbott commented:
For prototyping, I think you can add this overlay on "Why Zulip" -- with the individual /for/open-source type pages as the elements in the dropdown (rather than adding a new element).  That's the one that we'd be most likely in a position to merge if we were happy with the technical details of the implementation.

eeshangarg commented:
@timabbott: I am hoping to push the rest of the restructuring changes for the navigation tomorrow, so that should be even better! See [this Google sheet](https://docs.google.com/spreadsheets/d/1IIokZlIp4iFh-oaQmJBsQfHHCTn-JBHq8zZwg55aF1o/edit?usp=sharing) that @alya and I worked on! :)

eeshangarg commented:
Just updated this PR with the rest of the changes and a more detailed commit description. @timabbott, @alya: :)

Relevant screenshots:
<img width="1348" alt="2021-06-24 (1)" src="https://user-images.githubusercontent.com/7251823/123289306-ec7d4300-d4ea-11eb-98a4-352a6cdb02de.png">
<img width="1326" alt="2021-06-24" src="https://user-images.githubusercontent.com/7251823/123289323-f0a96080-d4ea-11eb-8fe9-9ed4ad52c5c8.png">

amanagr commented:
Looks good visually.

eeshangarg commented:
My latest commit adds a "For administrators" section to the "Resources" dropdown and adds horizontal separation of different dropdown subsections. @alya FYI :)

![2021-06-28](https://user-images.githubusercontent.com/7251823/123668564-ca9afd80-d815-11eb-9d78-5f9bba1ee80f.png)

eeshangarg commented:
@timabbott @alya This one is ready for a review. Adding backlinks to the ReadTheDocs pages is something that's on my plate right now but not a blocker for merging this. Currently, RTD links open in new tabs. Thanks!

eeshangarg commented:
ReadTheDocs has backlinks now as well.

Screenshot for RTD:
<img width="264" alt="readthedocsbacklink" src="https://user-images.githubusercontent.com/7251823/124028316-6cfde100-d9ce-11eb-9cc3-c7490dcd4f99.PNG">

Screenshot for help and api pages:
<img width="241" alt="helppagebacklink" src="https://user-images.githubusercontent.com/7251823/124028418-843cce80-d9ce-11eb-8270-e4113b76271e.PNG">

alya commented:
Should we label the backlink "Go to zulip.com" in both places?

alya commented:
Also, I think we should use sentence case, i.e. "Help center", not "Help Center".

alya commented:
Glad you got this working!

eeshangarg commented:
@alya: I addressed your comment about the sentence case. As for changing the backlink to say "Go to zulip.com" for the help/API pages as well, I am not sure if that is the right thing to do here. Since we are linking to `/` (root), for someone who is logged-in, it would lead to their Zulip realm, and for someone who is not logged in, it would lead to the main webpage, so the behavior is not exactly congruent with what "Go to zulip.com" would imply in this case. Thanks! :)

alya commented:
Ah, but I think we specifically want a button that goes to zulip.com, right? Otherwise, we haven't solved the problem of backlinks when someone navigates to /help from the top menus on zulip.com.

eeshangarg commented:
@alya @timabbott I addressed all feedback, this is ready for another review! :)

timabbott commented on templates/zerver/documentation_main.html:
```diff
@@ -10,7 +10,8 @@
         <div class="content">
-            <h1><a href="{{ doc_root }}" class="no-underline">Home</a></h1>
+            <h1 class="home-link-top"><a href="https://zulip.com/" class="no-underline">Go to Zulip.com</a></h1>
+            <h1><a href="{{ doc_root }}" class="no-underline">Home - {{ doc_root_title }}</a></h1>
```
I don't think this looks super good.  What if we just made the `Zulip |` part of the top-left index go to zulip.com?

![image](https://user-images.githubusercontent.com/2746074/124205012-317a1a00-da95-11eb-8d3e-29c5cd3d784c.png)

There's also a question of whether zulip.com is the right link on a self-hosted server, or whether we want it to be `href=/` (i.e. go to the root of the current server you're looking at).

alya replied:
What would `href=/` do, exactly? The goal is for it to be easy to get back to zulip.com if you navigate to help through the top menus.

eeshangarg replied:
@timabbott: I linked to `/` earlier but changed it after Alya suggested that we go straight to zulip.com.

@alya: Linking to `href=/` takes you to the root of the current server for self-hosted servers. For our zulip.com domain, linking to `href=/` would still take you to the main website, so it actually doesn't matter for our purposes. But I think I agree with Tim here, it makes sense to make the link work more predictably for self-hosted servers, so I'll change it to `href=/`. 

Thanks!

timabbott replied:
OK, after some discussion, I think we should just modify master in this spot by replacing "Home" with a "Zulip homepage" entry that links to zulip.com.  If we leave the "Zulip Help Center" link at the top as-is, that's a fine way to get back to the index for those who want that; not sure we need it duplicated in the sidebar.

(Also, I think this shouldn't have been marked as resolved, since there's substantial discussion that hadn't been read by others)

timabbott commented on zerver/views/documentation.py:
```diff
@@ -131,12 +131,14 @@ def get_context_data(self, **kwargs: Any) -> Dict[str, Any]:
         if self.path_template == "/zerver/help/%s.md":
             context["page_is_help_center"] = True
             context["doc_root"] = "/help/"
+            context["doc_root_title"] = "Help center"
```
Do we want to rename the "User documentation" top-level heading for the page to "Help center" as well?

alya replied:
Yeah, I like "Help center" as the canonical name for these pages that we should consistently use everywhere.

eeshangarg replied:
Agreed! :)

timabbott commented on templates/zerver/landing_nav.html:
```diff
@@ -16,26 +16,89 @@
+                            <a href="/help">Help center</a>
+                        </li>
+                        <li>
+                            <a href="https://zulip.readthedocs.io/en/latest/contributing/chat-zulip-org.html" target=”_blank”>
```
Our policy is generally that all links with `target=_blank` need `rel="noopener noreferer"` as well.

timabbott commented on templates/zerver/landing_nav.html:
```diff
@@ -16,26 +16,89 @@
+                    <div class="dropdown-column">
+                        <li class="nav-header">For administrators</li>
+                        <li>
+                            <a href="https://zulip.readthedocs.io/en/stable/production/index.html" target=”_blank”>
```
This is the wrong link; it should go here: https://zulip.readthedocs.io/en/stable/production/install.html.

timabbott commented on templates/zerver/landing_nav.html:
```diff
@@ -16,26 +16,89 @@
+                        <li class="nav-header">For administrators</li>
+                        <li>
+                            <a href="https://zulip.readthedocs.io/en/stable/production/index.html" target=”_blank”>
+                                Installing a Zulip server
```
Maybe "Install a Zulip server"?

alya replied:
I'm not sure -- it takes you to instructions for how to do it, not to the place where you actually do it, right? For instructions, "Installing" might be better.

eeshangarg replied:
I agree with @alya here. The "installing" phrasing is also in line with "getting" started and "setting up an organization". So a continuous verb makes more sense to me here. Thanks!

eeshangarg replied:
But if we are going to change this, we should probably change it everywhere except "getting started".

timabbott replied:
I guess this is probably fine for a "resources" area; we can always tweak after playing with it more.

timabbott commented on templates/zerver/landing_nav.html:
```diff
@@ -16,26 +16,89 @@
+                            <a href="https://zulip.readthedocs.io/en/stable/production/index.html" target=”_blank”>
+                                Installing a Zulip server
+                            </a>
+                        </li>
```
Do we want "Upgrade a Zulip server" to be right after the installation docs?  I think it is probably a nice touch to do so, especially since the after sysadmin should upgrade more often than install.

alya replied:
I like it!

alya commented:
@eeshangarg For convenience, could you update the spreadsheet with any changes to link targets and link names?

eeshangarg commented:
@alya, @timabbott  I updated the spreadsheet and also addressed all feedback. This is now ready for another review! :)

timabbott commented:
Great!  I merged the /help/ commit as the series ending with fc0035f050e3539f987905760864a687d97da269 after a few small tweaks (including a new small CSS commit), most notably making the top link "Zulip homepage" (and skipping the previously "Home" link altogether, since it's not obviously useful).

For the main navigation commit, the main problem is that it doesn't work on mobile web as well as tablets:
* Touching the headings should open them if you don't have a mouse to mouseover.
* If you're in a narrow screen, we should just show all the items with headings vertically stacked here, like we do with the footer: 
![image](https://user-images.githubusercontent.com/2746074/124645553-706aef80-de48-11eb-91f1-3b97f5f7a50b.png)

timabbott commented on templates/zerver/landing_nav.html:
```diff
@@ -16,26 +16,94 @@
+                            </a>
+                        </li>
+                        <li>
+                            <a href="https://github.com/zulip/zulip" target=”_blank” rel="noopener noreferer">GitHub</a>
```
Careful!  These are curly quotes, which I don't think we ever want to use in HTML.

eeshangarg replied:
Thanks for catching that one! Should be fixed now.

timabbott commented:
I merged the ReadTheDocs commit as ef6582edb8d6ac61f528e20f6b532f9dbca2e8dd, after the same "Go to Zulip => Zulip homepage" change.

alya commented:
Poking around the navigation a bit more, here are a few additional requests:

- [ ] Whenever we show the "Find accounts" link, show a "New organization" link just to the right of it. It should go to `/new`.
- [ ] Add "Find accounts" and "New organization" links to the /help header (in the upper right), with the same logic for when to show them/not show them as on the landing page.
- [ ] Rename the "Home" link in the little drop-down that shows up on the right when you're logged in to "Go to app".

Thanks!

eeshangarg commented:
@alya: I made the change about the reordering you requested. 

@timabbott: Starting playing around with the CSS on mobile and got to here (note that the dropdowns still need work):
<img width="350" alt="latest_mobile_nav" src="https://user-images.githubusercontent.com/7251823/124833544-e9f20300-df58-11eb-8933-b3305153370b.png">

eeshangarg commented:
@timabbott @alya So I spent a bit more time on this today and pushed two WIP commits, one with no toggle but with a scrollbar that looks like this:

<img width="284" alt="nav_with_scrollbar" src="https://user-images.githubusercontent.com/7251823/125348712-26e23f00-e337-11eb-8222-4599bd3b3de5.PNG">

And one with no scrollbar but toggles that show each section when hovered over. I am not sure how I feel about the side navigation having its own scrollbar on mobile. We could easily implement toggles here by disabling toggling on hovering and enabling click toggling for smaller screen widths. 

I would really appreciate some feedback on this. Personally, I think toggles seem a bit less cluttered and cleaner, but I am not opposed to the scrollbar (especially if we can make the scrollbar less ugly maybe?)

Thanks!

alya commented:
Hm, I'm a bit confused. Your screenshot looks like it has the scroll bar and toggles? Or is that just visually, and actually the menus are open all the time?

Anyhow, if you have the toggles working, I think that's great and we should go with that. The primary use case for these narrow windows is mobile, where there is no hover behavior. On mobile, I would expect not to have a visible scroll bar, except for when you're pulling up/down, the way you generally works on other apps. (I'm personally and Android user, but I assume the standard UI for this is the same on iOS?)

In narrow windows on desktop, I think it's reasonable to open the toggles on click as well, and to show a scroll bar when you can scroll. It should be rare for this to come up in any case.

eeshangarg commented:
@alya: Yes, that is just visually, the menus are open all the time. I just wanted to show the visual placement of the elements.

Thanks for the feedback, I'll just implement the clicky toggles with an optional scrollbar for content overflows. Thanks!

eeshangarg commented:
@alya This is ready for another review. All of your feedback so far has been addressed. Final screenshot:
<img width="285" alt="latest_mobild_nav_2" src="https://user-images.githubusercontent.com/7251823/125704640-e277af24-f491-4f9c-a498-5882efefac77.PNG">

alya commented:
Looking pretty good!

Can we make the subheadings (e.g. "for users") a little bigger? They are a bit hard to read at the moment.

eeshangarg commented:
@alya: Done! :)

<img width="282" alt="bigger_text_subheadings" src="https://user-images.githubusercontent.com/7251823/125706353-9fe03a46-61b1-4351-b702-d228e04c4cf8.PNG">

timabbott commented:
I made the following CSS changes, mainly to prevent unnecessary line-wrapping in various screen sizes:

```
diff --git a/static/styles/portico/landing_page.css b/static/styles/portico/landing_page.css
index 1de6134a18..7a28e62900 100644
--- a/static/styles/portico/landing_page.css
+++ b/static/styles/portico/landing_page.css
@@ -325,21 +325,28 @@ nav {
 
     .nav-dropdown {
         padding-bottom: 3px;
-
-        ul.product-options {
-            width: 125px;
+        ul {
+            width: 300px;
         }
 
-        ul.solutions-options {
-            width: 175px;
-        }
+        @media (width > 686px) {
+            ul.product-options {
+                width: 125px;
+            }
 
-        ul.resources-options {
-            width: 300px;
+            ul.solutions-options {
+                width: 175px;
+            }
 
-            div.dropdown-column {
-                width: 50%;
-                float: left;
+            ul.resources-options {
+                width: 350px;
+
+                div.dropdown-column {
+                    float: left;
+                    &:first-of-type {
+                        width: 40%;
+                    }
+                }
             }
         }
     }
@@ -3560,7 +3567,7 @@ nav {
             div.dropdown ul.dropdown-list {
                 position: relative;
                 box-shadow: none;
-                padding-left: 25px;
+                padding-left: 10px;
                 border: none;
 
                 li {
@@ -3581,7 +3588,6 @@ nav {
 
                         li.nav-header {
                             font-size: 1em;
-                            width: 175px;
                         }
                     }
                 }
@@ -3599,7 +3605,7 @@ nav {
     .portico-header .content > ul::before {
         content: "Zulip";
         display: block;
-        margin-top: 20px;
+        margin-top: 25px;
         padding-top: 5px;
         margin-left: 10px;
 
diff --git a/static/styles/portico/portico.css b/static/styles/portico/portico.css
index 6756a2ea21..1e1d37ab71 100644
--- a/static/styles/portico/portico.css
+++ b/static/styles/portico/portico.css
@@ -589,7 +589,7 @@ input.text-error {
             right: 0;
             margin: 10px 0 0 0;
             list-style-type: none;
-            width: 200px;
+            width: 300px;
             background-color: hsl(0, 0%, 100%);
             padding: 5px 0;
             border-radius: 4px;
```

timabbott commented:
Following those changes, I've merged this. However, I just noted an important bug that we still need to fix as a near-term follow-up: On tablets (E.g. 800px width), it's impossible to open the navigation (clicking the widgets with the a touch mouse doesn't work); can you investigate that?  Easily reproduced with the browser inspector.

I suspect the `window.matchMedia("(max-width: 686px)").matches` check in `static/js/portico/header.js` is the problem, but just deleting that causes some secondary problem.
